sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala

```diff
@@ -181,19 +181,35 @@ private case object OracleDialect extends JdbcDialect {
     if (limit > 0) s"WHERE rownum <= $limit" else ""
   }
 
+  override def getOffsetClause(offset: Integer): String = {
+    // Oracle doesn't support OFFSET clause.
+    // We can use rownum > n to skip some rows in the result set.
+    // Note: rn is an alias of rownum.
+    if (offset > 0) s"WHERE rn > $offset" else ""
+  }
+
   class OracleSQLQueryBuilder(dialect: JdbcDialect, options: JDBCOptions)
     extends JdbcSQLQueryBuilder(dialect, options) {
 
-    // TODO[SPARK-42289]: DS V2 pushdown could let JDBC dialect decide to push down offset
     override def build(): String = {
       val selectStmt = s"SELECT $columnList FROM ${options.tableOrQuery} $tableSampleClause" +
         s" $whereClause $groupByClause $orderByClause"
-      if (limit > 0) {
-        val limitClause = dialect.getLimitClause(limit)
-        options.prepareQuery + s"SELECT tab.* FROM ($selectStmt) tab $limitClause"
+      val finalSelectStmt = if (limit > 0) {
+        if (offset > 0) {
+          s"SELECT $columnList FROM (SELECT tab.*, rownum rn FROM ($selectStmt) tab)" +
```

Comment:
how about
```
SELECT * FROM ($selectStmt) tab WHERE rownum > ...
```

Reply:
Maybe it is a bug of Oracle.
If we using rownum directly, the result will incorrect.

Reply:
https://stackoverflow.com/questions/31186166/oracle-sql-filtering-by-rownum-not-returning-results-when-it-should

Let's mention the reason as well: the `rownum` is calculated when the value is returned.

Reply:
or we can use the new syntax in oracle 12+, which should be the widely used versions today.

Reply:
> or we can use the new syntax in oracle 12+, which should be the widely used versions today.

Could we upgrade the version of oracle in another PR if it is widely used.